File: Cryambly/Engine/Models/Characters/Attachments/Objects/AttachedObjectsCommons.cs
using System;
using System.Runtime.CompilerServices;
using CryCil.Engine.Logic;
using CryCil.Engine.Models.StaticObjects;
using CryCil.Engine.Rendering;
using CryCil.Engine.Rendering.Lighting;
using CryCil.Geometry;

namespace CryCil.Engine.Models.Characters.Attachments
{
	internal static class AttachedObjectsCommons
	{
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void ProcessAttachment(IntPtr handle, AttachmentSocket pIAttachment);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern BoundingBox GetAabb(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern float GetRadiusSqr(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern StaticObject GetIStatObj(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void SetIStatObj(IntPtr handle, StaticObject staticObject);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern Character GetICharacterInstance(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void SetICharacterInstance(IntPtr handle, Character character);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern AttachmentSkin GetIAttachmentSkin(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void SetIAttachmentSkin(IntPtr handle, AttachmentSkin skin);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern string GetObjectFilePath(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern Material GetBaseMaterial(IntPtr handle, uint nLod);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void SetReplacementMaterial(IntPtr handle, Material pMaterial, uint nLod);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern Material GetReplacementMaterial(IntPtr handle, uint nLod);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void Release(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void SetEntityId(IntPtr handle, EntityId id);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern EntityId GetEntityId(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern void LoadLight(IntPtr handle, ref LightProperties light);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern LightSource GetLightSource(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern ParticleEmitter GetEmitter(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern IntPtr LoadEffectAttachment(string effectName, ref Vector3 offset, ref Vector3 dir, float scale);
		[MethodImpl(MethodImplOptions.InternalCall)]
		internal static extern IntPtr CreateEffectAttachment(ParticleEffect pParticleEffect, ref Vector3 offset, ref Vector3 dir, float scale);
	}
}

File: MonoInterface/Interops/AttachedObjectsCommons.cpp
#include "stdafx.h"

#include "AttachedObjectsCommons.h"
#include <IAttachment.h>

void AttachedObjectsCommonsInterop::OnRunTimeInitialized()
{
	REGISTER_METHOD(ProcessAttachment);
	REGISTER_METHOD(GetAabb);
	REGISTER_METHOD(GetRadiusSqr);
	REGISTER_METHOD(GetIStatObj);
	REGISTER_METHOD(SetIStatObj);
	REGISTER_METHOD(GetICharacterInstance);
	REGISTER_METHOD(SetICharacterInstance);
	REGISTER_METHOD(GetIAttachmentSkin);
	REGISTER_METHOD(SetIAttachmentSkin);
	REGISTER_METHOD(GetObjectFilePath);
	REGISTER_METHOD(GetBaseMaterial);
	REGISTER_METHOD(SetReplacementMaterial);
	REGISTER_METHOD(GetReplacementMaterial);
	REGISTER_METHOD(Release);
	REGISTER_METHOD(SetEntityId);
	REGISTER_METHOD(GetEntityId);
	REGISTER_METHOD(LoadLight);
	REGISTER_METHOD(GetLightSource);
	REGISTER_METHOD(GetEmitter);
	REGISTER_METHOD(LoadEffectAttachment);
	REGISTER_METHOD(CreateEffectAttachment);
}

void AttachedObjectsCommonsInterop::ProcessAttachment(IAttachmentObject *handle, IAttachment *pIAttachment)
{
	handle->ProcessAttachment(pIAttachment);
}

AABB AttachedObjectsCommonsInterop::GetAabb(IAttachmentObject *handle)
{
	return handle->GetAABB();
}

float AttachedObjectsCommonsInterop::GetRadiusSqr(IAttachmentObject *handle)
{
	return handle->GetRadiusSqr();
}

IStatObj *AttachedObjectsCommonsInterop::GetIStatObj(CCGFAttachment *handle)
{
	return handle->pObj;
}

void AttachedObjectsCommonsInterop::SetIStatObj(CCGFAttachment *handle, IStatObj *staticObject)
{
	handle->pObj = staticObject;
}

ICharacterInstance *AttachedObjectsCommonsInterop::GetICharacterInstance(CSKELAttachment *handle)
{
	return handle->m_pCharInstance;
}

void AttachedObjectsCommonsInterop::SetICharacterInstance(CSKELAttachment *handle, ICharacterInstance *character)
{
	handle->m_pCharInstance = character;
}

IAttachmentSkin *AttachedObjectsCommonsInterop::GetIAttachmentSkin(CSKINAttachment *handle)
{
	return handle->m_pIAttachmentSkin;
}

void AttachedObjectsCommonsInterop::SetIAttachmentSkin(CSKINAttachment *handle, IAttachmentSkin *skin)
{
	handle->m_pIAttachmentSkin = skin;
}

mono::string AttachedObjectsCommonsInterop::GetObjectFilePath(IAttachmentObject *handle)
{
	return ToMonoString(handle->GetObjectFilePath());
}

IMaterial *AttachedObjectsCommonsInterop::GetBaseMaterial(IAttachmentObject *handle, uint nLod)
{
	return handle->GetBaseMaterial(nLod);
}

void AttachedObjectsCommonsInterop::SetReplacementMaterial(IAttachmentObject *handle, IMaterial *pMaterial, uint nLod)
{
	handle->SetReplacementMaterial(pMaterial, nLod);
}

IMaterial *AttachedObjectsCommonsInterop::GetReplacementMaterial(IAttachmentObject *handle, uint nLod)
{
	return handle->GetReplacementMaterial(nLod);
}

void AttachedObjectsCommonsInterop::Release(IAttachmentObject *handle)
{
	handle->Release();
}

void AttachedObjectsCommonsInterop::SetEntityId(CEntityAttachment *handle, EntityId id)
{
	handle->SetEntityId(id);
}

EntityId AttachedObjectsCommonsInterop::GetEntityId(CEntityAttachment *handle)
{
	return handle->GetEntityId();
}

void AttachedObjectsCommonsInterop::LoadLight(CLightAttachment *handle, const CDLight &light)
{
	handle->LoadLight(light);
}

ILightSource *AttachedObjectsCommonsInterop::GetLightSource(CLightAttachment *handle)
{
	return handle->GetLightSource();
}

IParticleEmitter *AttachedObjectsCommonsInterop::GetEmitter(CEffectAttachment *handle)
{
	return handle->GetEmitter();
}

CEffectAttachment *AttachedObjectsCommonsInterop::LoadEffectAttachment(mono::string effectName, const Vec3 &offset,
																	   const Vec3 &dir, float scale)
{
	return new CEffectAttachment(NtText(effectName), offset, dir, scale);
}

CEffectAttachment *AttachedObjectsCommonsInterop::CreateEffectAttachment(IParticleEffect *pParticleEffect,
																		 const Vec3 &offset, const Vec3 &dir, float scale)
{
	return new CEffectAttachment(pParticleEffect, offset, dir, scale);
}
